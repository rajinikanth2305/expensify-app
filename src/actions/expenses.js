
import uuid from "uuid"
import database from '../firebase/firebase';


//component calls action generator
//action generator returns object
//component dispatches object
//redux store changes

//component calls action generator
//action generator returns function
//component dispatches function
//function runs he ability to dispatch other actions and do whatever it wants




export const addExpense=(expense)=>(
    {
       type:"ADD_EXPENSE",
       expense
    }
)


export const startAddExpense = (expenseData = {}) => {
    return (dispatch,getState) => {
      const uid=getState().auth.uid;
      const {
        description = '',
        note = '',
        amount = 0,
        createAt = 0
      } = expenseData;
      const expense = { description, note, amount, createAt };
  
     return database.ref(`users/${uid}/expenses`).push(expense).then((ref) => {
        dispatch(addExpense({
          id: ref.key,
          ...expense
        }));
      });
    };
  };
  
//setup remove expense that we can dispatch to firebase
export const startRemoveExpense=({id}={})=>
{
  console.log(id);
  return (dispatch,getState)=>
  {
    const uid=getState().auth.uid;
    return database.ref(`users/${uid}/expenses/${id}`).remove().then(()=>
    {
      dispatch(removeExpense({id}))
    }).catch((e)=>
    {

    })
  }
}

//we have to return keyword fro database.ref because for test case
//set up edit expense that we can dispatch to firebase


//expenses/${actions[0].expense.id}
//remove expense


export const removeExpense=({id}={})=>(
{
    type:"REMOVE_EXPENSE",
    id

}
)
//edit expense


export const editExpense=(id,updates)=>(
    {
        type:"EDIT_EXPENSE",
        id,
        updates
    }
)
export const startEditExpense=(id,updates)=>
{
  return (dispatch,getState)=>
  {
    const uid=getState().auth.uid;
    return database.ref(`users/${uid}/expenses/${id}`).update(updates).then(()=>
    {
      dispatch(editExpense(id,updates))
    })
  }
}

//SET_EXPENSES
export const setExpenses=(expenses)=>(
  {
    type:"SET_EXPENSES",
    expenses
  }
)

//fetch all expenses data once 
//pass that data tothe array
//dispatch setExpenses
//export const startSetExpenses

export const startSetExpenses=()=>
{
  return (dispatch,getState)=>
  {
    const uid=getState().auth.uid;
    return database.ref(`users/${uid}/expenses`).once("value").then((snapshot)=>
    {
      const expenses=[];
      snapshot.forEach((childsnapshot)=>
      {
        expenses.push({
          id:childsnapshot.key,
          ...childsnapshot.val()
        })
      })
      dispatch(setExpenses(expenses))
    })
  }
}