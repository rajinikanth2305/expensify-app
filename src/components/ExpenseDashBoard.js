import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListItem from "./ExpenseListItem"
import ExpenseListFilters from "./ExpenseListFilter"
import ViewTotalExpenses from './ExpensesSummary'
const ExpenseDashboardPage = () => (
  <div>
    <ViewTotalExpenses />
    <ExpenseListFilters />
    <ExpenseList />
  </div>
);

export default ExpenseDashboardPage;
