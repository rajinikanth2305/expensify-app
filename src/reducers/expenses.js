//expenses reducer
import database from "../firebase/firebase"
const expensesReducerDefaultSate=[];

database.ref("expenses").once("value").then((snapshot)=>
{

    snapshot.forEach((childsnapshot)=>
    {
        expensesReducerDefaultSate.push(
            {
               id:childsnapshot.key,
               ...childsnapshot.val()
            }
        )
    })

})


export default (state=expensesReducerDefaultSate,action)=>
{
    switch(action.type)
    {
        case "ADD_EXPENSE":
            return [...state,action.expense]
        
        case "REMOVE_EXPENSE":
            return state.filter(({id})=>id!=action.id);
        case "EDIT_EXPENSE":
            return state.map((expense)=>
            {
                if(expense.id===action.id)
                {
                    return {...expense,...action.updates}
                }
                else{
                    return expense
                }

            })
        case "SET_EXPENSES":
            return action.expenses

        
            default:
            return state;
    }
}