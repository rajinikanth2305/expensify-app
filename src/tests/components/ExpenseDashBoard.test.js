import React from 'react';
import { shallow } from 'enzyme';
import expenses from '../fixtures/expenses';
import ExpenseDashBoard from '../../components/ExpenseDashBoard';
test("should render expense dashboard as exactly",()=>
{
    const wrapper=shallow(<ExpenseDashBoard />);
    expect(wrapper).toMatchSnapshot();


})