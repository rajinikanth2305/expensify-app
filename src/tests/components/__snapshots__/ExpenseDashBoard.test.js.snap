// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense dashboard as exactly 1`] = `
<div>
  <Connect(ViewTotalExpenses) />
  <Connect(ExpenseListFilters) />
  <Connect(ExpenseList) />
</div>
`;
